### ISSUE TITLE
Assertion failed: data already exists for generation 22

### ISSUE THREAD

**Issue by oliver-phet** [phetsims/natural-selection#165]
I just started looking at the sim again for https://github.com/phetsims/natural-selection/issues/155, trying to replicate the results from yesterday [using the same steps](https://github.com/phetsims/natural-selection/issues/153#issuecomment-668722992).
I am able to inconsistently trigger an assertion failure. Using latest on phettest `?ea&brand=phet&secondsPerGeneration=1&log` These are the actions I'm taking:

1. Console Open
2. long teeth recessive, limited food, add a mate, hold fast forward
3. Either the bunnies take over or many generations pass with the population stable
4. click reset all button
5. long teeth recessive, limited food, add a mate, hold fast forward

<img width="857" alt="image" src="https://user-images.githubusercontent.com/5090275/89466087-6ddfb180-d730-11ea-93f7-189cfd7acc20.png">
2nd time reproducing I was able to capture log:
[log2.txt](https://github.com/phetsims/natural-selection/files/5031396/log2.txt)

**Comment by oliver-phet** [comment]
This seems to be the relevant part of the log:
```
[Log] ====== Generation 21 ====== (initialize-globals.js, line 636)
[Log] 1 bunnies died of old age (initialize-globals.js, line 636)
[Log] 12 dead bunnies pruned (initialize-globals.js, line 636)
[Log] using bunnyRestRange=[6,10] for population=250 (initialize-globals.js, line 636)
[Log] 284 bunnies were born (initialize-globals.js, line 636)
[Log] Assertion failed: data already exists for generation 22 (assert.js, line 21)
[Log] Assertion failed: reentry detected, value=23.108, oldValue=23 (assert.js, line 21)
[Log] Assertion failed: reentry detected, value=23.144000000000002, oldValue=23.108 (assert.js, line 21)
[Log] Assertion failed: reentry detected, value=23.212000000000003, oldValue=23.144000000000002 (assert.js, line 21)

```

**Comment by pixelzoom** [comment]
@oliver-phet I need to know what you're running.  A dev version? phet-test?....

**Comment by oliver-phet** [comment]
Using latest on phettest `?ea&brand=phet&secondsPerGeneration=1&log`
MacOS 10.15.6
Safari Version 13.1.2 (15609.3.5.1.3)

**Comment by pixelzoom** [comment]
I can't reproduce with my local copy on master. I'll try doing a pull-all.sh to see if there might be a problem in common code that snuck in, but I doubt it.

Using `secondsPerGeneration=1` with fast-forward is definitely a problem.  It causes important "slices" of the clock to be skipped over.    But I'm going to address that over in #158.

The relevant assertion here is:
> [Log] Assertion failed: data already exists for generation 22 (assert.js, line 21)

All of the others ("reentry detected") are symptomatic of the sim trying to continue to run in a broken state.

**Comment by pixelzoom** [comment]
I still can't reproduce in my local copy after a pull-all.sh.

**Comment by pixelzoom** [comment]
I still haven't been able to reproduce this, and the Food model has been revised. 

@oliver-phet please attempt to reproduce this in phet-test or [1.1.0-dev.4](https://phet-dev.colorado.edu/html/natural-selection/1.1.0-dev.4/phet/natural-selection_en_phet.html).

**Comment by pixelzoom** [comment]
Hey, I finally did reproduce this, so assigning back to me.  I guess it must happen intermittently.   In my case, the error was:

```
Uncaught Error: Assertion failed: data already exists for generation 68
    at window.assertions.assertFunction (assert.js:22)
    at ProportionsModel.recordEndCounts (ProportionsModel.js:233)
    at NaturalSelectionModel.js:179
    at TinyEmitter.emit (TinyEmitter.js:69)
    at DerivedProperty._notifyListeners (Property.js:273)
    at DerivedProperty.set (Property.js:171)
    at listener (DerivedProperty.js:73)
    at TinyEmitter.emit (TinyEmitter.js:69)
    at DerivedProperty._notifyListeners (Property.js:273)
    at DerivedProperty.set (Property.js:171)
```

**Comment by pixelzoom** [comment]
Here's the log just before the above failure.  What's interesting here is that we should have been recording end counts for generation 67, not 68.

```
====== Generation 67 ======
initialize-globals.js:636 0 bunnies died of old age
initialize-globals.js:636 8 dead bunnies pruned
initialize-globals.js:636 mating 4 bunnies
initialize-globals.js:636 8 bunnies were born
initialize-globals.js:636 Applying food: 12 short teeth, 0 long teeth, limited=true, tough=false
initialize-globals.js:636 8 of 12 short teeth died of starvation
initialize-globals.js:636 0 of 0 long teeth died of starvation
assert.js:21 Assertion failed: data already exists for generation 68
```

**Comment by pixelzoom** [comment]
I the above commit, I changed the assertion message to be more informative, and I'm printing the generation number as soon as it changes.

It took me awhile, but I finally made the error happen again. Here the log just before, and the stack trace:

```
initialize-globals.js:636 ====== Generation 728 ======
initialize-globals.js:636 live bunnies = 4
initialize-globals.js:636 dead bunnies = 120
initialize-globals.js:636 recessive mutants = 0
initialize-globals.js:636 0 bunnies died of old age
initialize-globals.js:636 8 dead bunnies pruned
initialize-globals.js:636 mating 4 bunnies
initialize-globals.js:636 8 bunnies were born
initialize-globals.js:636 ====== Generation 730 ======
assert.js:21 Assertion failed: previousCounts is out of sync, length=728, generation=729
assert.js:22 Uncaught Error: Assertion failed: previousCounts is out of sync, length=728, generation=729
    at window.assertions.assertFunction (assert.js:22)
    at ProportionsModel.recordEndCounts (ProportionsModel.js:233)
    at NaturalSelectionModel.js:181
    at TinyEmitter.emit (TinyEmitter.js:69)
    at DerivedProperty._notifyListeners (Property.js:273)
    at DerivedProperty.set (Property.js:171)
    at listener (DerivedProperty.js:73)
    at TinyEmitter.emit (TinyEmitter.js:69)
    at DerivedProperty._notifyListeners (Property.js:273)
    at DerivedProperty.set (Property.js:171)
```

Notice that it skips from "Generation 728" to "Generation 730".   So what happened is we completely skipped over generation 729.   This is probably because I'm running with `secondsPerGeneration=0.25`, which is equivalent to what @oliver-phet was doing (`secondsPerGeneration=1` and fast-forward 4x).

**Comment by pixelzoom** [comment]
I the above commit, I added an assertion that generation should increase by 1.   That results in:

```
====== Generation 305 ======
initialize-globals.js:636 live bunnies = 3
initialize-globals.js:636 dead bunnies = 129
initialize-globals.js:636 recessive mutants = 0
initialize-globals.js:636 0 bunnies died of old age
initialize-globals.js:636 8 dead bunnies pruned
initialize-globals.js:636 mating 3 bunnies
initialize-globals.js:636 4 bunnies were born
initialize-globals.js:636 Applying food: 7 short teeth, 0 long teeth, limited=true, tough=false
initialize-globals.js:636 5 of 7 short teeth died of starvation
initialize-globals.js:636 0 of 0 long teeth died of starvation
assert.js:21 Assertion failed: skipped a generation, currentGeneration=307, previousGeneration=305
assert.js:22 Uncaught Error: Assertion failed: skipped a generation, currentGeneration=307, previousGeneration=305
    at window.assertions.assertFunction (assert.js:22)
    at GenerationClock.js:78
    at TinyEmitter.emit (TinyEmitter.js:69)
    at DerivedProperty._notifyListeners (Property.js:273)
    at DerivedProperty.set (Property.js:171)
    at listener (DerivedProperty.js:73)
    at TinyEmitter.emit (TinyEmitter.js:69)
    at DerivedProperty._notifyListeners (Property.js:273)
    at DerivedProperty.set (Property.js:171)
    at listener (DerivedProperty.js:73)
```

I've also noticed that if I run with `secondsPerGeneration` <= 0.2, the sim will stop and hang at generation 42.  There's no error in the console, everything just stops.   I waited for awhile, and there was no stack overflow.

So there are definitely multiple problems with running the clock at extreme speeds.  This will not be a problem in practice.  But I'm going to pursue it because it's an important development and testing tool.

**Comment by pixelzoom** [comment]
In the above commit, I limited dt such that we have a minimum of 10 steps per cycle of the generation clock.  This will allow us to run the clock at very fast speeds without skipping a generation.  

This did not address the problem of the sim hanging when `secondsPerGeneration` <= 0.2.  I'm still investigating, but I may end up requiring `secondsPerGeneration` to be > 0.2.

**Comment by pixelzoom** [comment]
What's currently happening with `secondsPerGeneration=0.2` is very odd.  In GenerationClock.step, `this.timeProperty.value` is being repeatedly set to 8.6, instead of advancing.  So as far as the rest of the sim is concerned, the generation clock has effectively stopped.

**Comment by pixelzoom** [comment]
I'm glad that I pursue this.  It turned out to be a floating-point error in the conversion to elapsed time to generations. 

One would think that 8.6 / 0.2 = 43, but in JavaScript it's 42.99999999999999.  So when we hit generation 43, we got stuck and couldn't advance.   I had compensated for this in one place by adding a small value (0.0001), but it needed to be done in 2 places.  It's not factored out into `function timeToGenerations`.
